File: AlgorithmForCompetition/Graph/Topology.js
const indegree = [0,1,1,2,1,2,1];
const graph = [[1,4],[2,5],[3],[6],[5],[3],[]];
const que = [];
const result = [];

//-----------------------------[input]

que.push(0);

while(que.length > 0) {
  const now = que.shift();
  result.push(now);

  for(let i=0;i<graph[now].length;i++) {
    let degree = graph[now][i];

    indegree[degree] --;
    if(indegree[degree] === 0) {
      que.push(degree);
    }
  }
}

console.log(result);

File: RestoreAlgorithm/Graph/Topology.cpp
#include <iostream>
#include <string>
#include <vector>
#include <algorithm>
#include <queue>

using namespace std;

int main()
{
  ios_base ::sync_with_stdio(false);
  cin.tie(NULL);

  int verNum, edgeNum;
  vector<vector<int>> graph(verNum + 1);
  vector<int> indegree(verNum + 1, 0);
  vector<int> result;
  queue<int> que;

  cin >> verNum >> edgeNum;

  for (int i = 0; i < edgeNum; i++)
  {
    int a, b;
    cin >> a >> b;

    graph[a].push_back(b);
    indegree[b]++;
  }

  for (int i = 1; i <= verNum; i++)
  {
    if (indegree[i] == 0)
    {
      que.push(i);
    }
  }

  while (!que.empty())
  {
    int now = que.front();
    que.pop();

    result.push_back(now);

    for (int i = 0; i < graph[now].size(); i++)
    {
      int next = graph[now][i];
      indegree[next]--;

      if (indegree[next] == 0)
      {
        que.push(next);
      }
    }
  }

  for (int i = 0; i < result.size(); i++)
  {
    cout << result[i] << " ";
  }
  cout << endl;
}
